classes/helper.php

```diff
@@ -34,23 +36,105 @@ class helper {
      * Get columns to search for in a table.
      *
      * @param string $table The table to search.
-     * @param string $column The column to search.
+     * @param array $searchingcolumns The columns to search.
      * @return array The columns to search.
      */
-    public static function get_columns(string $table, string $column = ''): array {
+    public static function get_columns(string $table, array $searchingcolumns = []): array {
         global $DB;
         $columns = $DB->get_columns($table);
 
-        if ($column !== self::ALL_COLUMNS) {
-            // Only search the specified column.
-            $columns = array_filter($columns, function($col) use ($column) {
-                return $col->name == $column;
-            });
+        // Make sure the table has id field.
+        $hasid = false;
+        foreach ($columns as $col) {
+            if ($col->name == 'id') {
+                $hasid = true;
+                break;
+            }
         }
 
+        // Do not search if the table does not have id field.
+        if (!$hasid) {
+            return [];
+        }
+
+        // Only search the specified columns.
+        foreach ($searchingcolumns as $column) {
+            if ($column !== self::ALL_COLUMNS) {
+                $columns = array_filter($columns, function($col) use ($column) {
+                    return $col->name == $column;
+                });
+            }
+        }
+
+        // Check if we need to skip some columns.
+        $columns = array_filter($columns, function($col) use ($table) {
+            return db_should_replace($table, $col->name);
+        });
+
         return $columns;
     }
 
+    /**
+     * Build searching list
+     *
+     * @param string $tables A comma separated list of tables and columns to search.
+     * @return array
+     */
+    public static function build_searching_list(string $tables = ''): array {
```

Comment:
This is returning lots of columns that are not text. This should never return things like id columns, or lastmod, or dates.

* [ ] It should only return varchar / text columns

* [ ] On top of that we can exclude some know ones like if we have intro and introformat we can safely ignore introformat.

* [ ] a good heuristic is if the search text is 100 chars long, and the column we are searching for is a varchar 20 then we can skip it

Reply:
tool_replace already does some of this logic so we can copy logic from there

Reply:
Hi @brendanheywood ,

I have added the logic